### PR TITLE
Monolithic fiber DRC memory usage reduction update, using IDEA o1 v03

### PR DESCRIPTION
**Pull request by swkim95**
BEGINRELEASENOTES
- Add DRC geometry construction code under `detector/calorimeter/dual-readout` directory
- Add .xml compact files under `FCCee/IDEA/compact/IDEA_o1_v03` directory
- Add custom SD action, output file, fast simulation (boosting optical photon transportation) for Monolithic fiber DRC under `plugin` directory
- Fixed CMakeLists to compile all above

ENDRELEASENOTES

Purpose
- PR for adding IDEA monolithic fiber dual-readout calorimeter (DRC) detector geometry

Tests
- Tested compiling k4geo
- Tested running simulation using ddsim

### CODE REVIEW

File: detectorSegmentations/src/GridDRcalo_k4geo.cpp
Comment (SanghyunKo) [inline detectorSegmentations/src/GridDRcalo_k4geo.cpp:97]: since we now copy LHS from RHS, `numEta(vID)` will always return positive `numEta`. To discriminate LHS towers from RHS towers, we can make use of `globalPosition` so that if `globalPosition.z() >= 0` then return `numEta(vID)` and if it is negative then return `-numEta(vID)-1`

**Review by SanghyunKo — CHANGES_REQUESTED** [review]
The first round of comments (note that assembly box should also be changed - which I know that is ongoing)

**Review by SanghyunKo — APPROVED** [review]

**Comment by SanghyunKo** [comment]
@BrieucF PR looks good to me. Would you mind reviewing it? Otherwise, I think we can merge this for the sake of material budget studies.

File: FCCee/IDEA/compact/IDEA_o1_v03/IDEA_o1_v03.xml
Comment (BrieucF) [inline FCCee/IDEA/compact/IDEA_o1_v03/IDEA_o1_v03.xml:45]: why is this commented out?

**Comment by BrieucF** [comment]
Thanks @SanghyunKo and @swkim95 ! Alvaro will do the review as soon as possible. I'd just ask for now to add a description of the detector builder here `detector/calorimeter/README.md` and a "change log" for IDEA_o1_v03 here: https://github.com/key4hep/k4geo/tree/main/FCCee/IDEA/compact#idea_o1_v03

File: FCCee/IDEA/compact/IDEA_o1_v03/IDEA_o1_v03.xml
Comment (swkim95) [reply FCCee/IDEA/compact/IDEA_o1_v03/IDEA_o1_v03.xml:45]: Oh I forgot to place it back. I'll update it.

I commented it out to test running DRC simulation.
I did it because when compiling the k4geo I got this warning message (even without any new code, freshly cloned from k4geo main branch) 

`CMake Warning at CMakeLists.txt:106 (message):
   Subdetector DriftChamber_o1_v02.cpp will not be built because header file
   was not found`

and if I don't comment drift chamber out and start simulation (even w/o implementation of DRC) I got this error

```
****************************************************
*  A runtime error has occured :
*    Compact: Failed to execute subdetector creation plugin. 		No factory with name Create(DriftChamber_o1_v02_T) for type DriftChamber_o1_v02_T found.
		Please check library load path and/or plugin factory name.
*  the program will have to be terminated - sorry.
****************************************************
```

So I just commented out drift chamber and ran the ddsim with DRC.

Comment (BrieucF) [reply FCCee/IDEA/compact/IDEA_o1_v03/IDEA_o1_v03.xml:45]: Hi Sungwon, yes this is because the drift chamber builder is not yet included in the stable stack (last one is from March), we currently run this detector with the nightlies (`source /cvmfs/sw-nightlies.hsf.org/key4hep/setup.sh`)

File: plugins/Geant4Output2EDM4hep_DRC.cpp
Comment (atolosadelgado) [inline plugins/Geant4Output2EDM4hep_DRC.cpp:453]: `getParameter` returns an optional, to access the variable we have to append `.value()` as in the following:
```
    eventWeight = m_frame.getParameter<double>("EventWeights").value();
    ```

File: detector/calorimeter/dual-readout/src/DRconstructor.cpp
Comment (andresailer) [inline detector/calorimeter/dual-readout/src/DRconstructor.cpp:243]: unit!

File: detector/calorimeter/dual-readout/src/DRconstructor.cpp
Comment (andresailer) [inline detector/calorimeter/dual-readout/src/DRconstructor.cpp:240]: Unit

File: detector/calorimeter/dual-readout/src/DRconstructor.cpp
Comment (andresailer) [inline detector/calorimeter/dual-readout/src/DRconstructor.cpp:245]: unit

File: detectorSegmentations/src/GridDRcalo_k4geo.cpp
Comment (andresailer) [inline detectorSegmentations/src/GridDRcalo_k4geo.cpp:189]: This whole function needs comments.

File: plugins/FiberDRCaloSDAction.cpp
Comment (andresailer) [inline plugins/FiberDRCaloSDAction.cpp:91]: Unit

File: plugins/FiberDRCaloSDAction.cpp
Comment (andresailer) [inline plugins/FiberDRCaloSDAction.cpp:121]: units?

File: plugins/FiberDRCaloSDAction.cpp
Comment (andresailer) [inline plugins/FiberDRCaloSDAction.cpp:89]: Just rely on the internal unit and don't bother converting yourself?

File: plugins/FiberDRCaloSDAction.cpp
Comment (andresailer) [inline plugins/FiberDRCaloSDAction.cpp:87]: Is this a pass or deny filter? Comment!

File: plugins/FiberDRCaloSDAction.cpp
Comment (andresailer) [inline plugins/FiberDRCaloSDAction.cpp:152]: I still can't tell if this is a pass or deny filter. Does passing the filter mean rejection or acceptance?

File: detector/calorimeter/dual-readout/include/DRconstructor.h
Comment (andresailer) [inline detector/calorimeter/dual-readout/include/DRconstructor.h:38]: unit!

File: detectorSegmentations/src/GridDRcalo_k4geo.cpp
Comment (andresailer) [inline detectorSegmentations/src/GridDRcalo_k4geo.cpp:48]: `delete nullptr` is essentially a no-op, so the check for nullptr-ness is unnecessary.
What might be useful is setting, e.g.,  `fParamBarrel=nulltpr` in the constructor to prevent the attempt from `delete`ing random memory.

**Review by andresailer — COMMENTED** [review]
I can only give rather superficial comments, because I cannot understand the details of the geometry and implementation.

I would say that all of this needs a lot more comments in the code to understand the intent and for maintainability.

I probably missed some unit less numbers as well.

File: detector/calorimeter/dual-readout/src/DRconstructor.cpp
Comment (SanghyunKo) [reply detector/calorimeter/dual-readout/src/DRconstructor.cpp:63]: Thanks for catching this... Indeed this (and other unitless numbers) should be fixed.

File: plugins/Geant4Output2EDM4hep_DRC.cpp
Comment (swkim95) [reply plugins/Geant4Output2EDM4hep_DRC.cpp:453]: If I add .value(), then I get this error when compiling

```
/u/user/swkim/KEY4HEP/for_k4geo_pr_tmp/k4geo_memUpdateVer/k4geo/plugins/Geant4Output2EDM4hep_DRC.cpp: In member function 'virtual void dd4hep::sim::Geant4Output2EDM4hep_DRC::saveEvent(dd4hep::sim::Geant4OutputAction::OutputContext<G4Event>&)':
/u/user/swkim/KEY4HEP/for_k4geo_pr_tmp/k4geo_memUpdateVer/k4geo/plugins/Geant4Output2EDM4hep_DRC.cpp:453:64: error: request for member 'value' in '((dd4hep::sim::Geant4Output2EDM4hep_DRC*)this)->dd4hep::sim::Geant4Output2EDM4hep_DRC::m_frame.podio::Frame::getParameter<double>(std::__cxx11::basic_string<char>(((const char*)"EventWeights"), std::allocator<char>()))', which is of non-class type 'podio::GenericDataReturnType<double>' {aka 'double'}
  453 |     eventWeight = m_frame.getParameter<double>("EventWeights").value();
```

Is this expected?

Comment (atolosadelgado) [reply plugins/Geant4Output2EDM4hep_DRC.cpp:453]: I would say it is not expected. I was using key4hep nightlies for Alma9, and I got no error. What key4hep release (stable, nightlies) and what operative system are you using? 
Maybe @jmcarcell can help here :)

File: plugins/Geant4Output2EDM4hep_DRC.cpp
Comment (tmadlener) [inline plugins/Geant4Output2EDM4hep_DRC.cpp:40]:
```suggestion
#if PODIO_BUILD_VERSON >= PODIO_VERSION(0, 99, 0)
```
Easier to read and will definitely do what you want

File: plugins/Geant4Output2EDM4hep_DRC.cpp
Comment (tmadlener) [inline plugins/Geant4Output2EDM4hep_DRC.cpp:121]:
```suggestion
#if PODIO_BUILD_VERSION > PODIO_VERSION(0, 16, 2)
```

File: plugins/Geant4Output2EDM4hep_DRC.cpp
Comment (tmadlener) [inline plugins/Geant4Output2EDM4hep_DRC.cpp:143]:
```suggestion
#if PODIO_BUILD_VERSION > PODIO_VERSION(0, 16, 2)
```

File: plugins/Geant4Output2EDM4hep_DRC.cpp
Comment (tmadlener) [inline plugins/Geant4Output2EDM4hep_DRC.cpp:240]: see above

File: plugins/Geant4Output2EDM4hep_DRC.cpp
Comment (tmadlener) [reply plugins/Geant4Output2EDM4hep_DRC.cpp:453]: I am not entirely sure why you need a copy of the Geant4 output action here. In any case, there have been a few changes to the DD4hep one that should make all of this work transparently with all versions of podio and EDM4hep

- https://github.com/AIDASoft/DD4hep/pull/1264
- https://github.com/AIDASoft/DD4hep/pull/1252
- https://github.com/AIDASoft/DD4hep/pull/1251
- https://github.com/AIDASoft/DD4hep/pull/1198

Comment (jmcarcell) [reply plugins/Geant4Output2EDM4hep_DRC.cpp:453]: It's probably because the code is being compiled on top of the latest release and not the nightlies. I would recommend compiling with the nightlies first and then doing like https://github.com/AIDASoft/DD4hep/pull/1264 and others if needed to work on the release (or alternatively wait for the next release that will work if it works with the nightlies).

**Comment by BrieucF** [comment]
I see some warnings when compiling this version. Can you try to fix them?

```
/afs/cern.ch/user/b/brfranco/work/public/check_monolithic_dr/k4geo/detector/calorimeter/dual-readout/src/DRconstructor.cpp: In member function 'void ddDRcalo::DRconstructor::placeAssembly(xml_comp_t&, xml_comp_t&, dd4hep::DDSegmentation::DRparamBase_k4geo*, dd4hep::Volume&, dd4hep::Volume&, dd4hep::Volume&, int, int, bool)':
/afs/cern.ch/user/b/brfranco/work/public/check_monolithic_dr/k4geo/detector/calorimeter/dual-readout/src/DRconstructor.cpp:118:57: warning: unused parameter 'x_theta' [-Wunused-parameter]
  118 | void ddDRcalo::DRconstructor::placeAssembly(xml_comp_t& x_theta, xml_comp_t& x_wafer, dd4hep::DDSegmentation::DRparamBase_k4geo* param,
      |                                             ~~~~~~~~~~~~^~~~~~~
/afs/cern.ch/user/b/brfranco/work/public/check_monolithic_dr/k4geo/detector/calorimeter/dual-readout/src/DRconstructor.cpp:118:78: warning: unused parameter 'x_wafer' [-Wunused-parameter]
  118 | void ddDRcalo::DRconstructor::placeAssembly(xml_comp_t& x_theta, xml_comp_t& x_wafer, dd4hep::DDSegmentation::DRparamBase_k4geo* param,
      |                                                                  ~~~~~~~~~~~~^~~~~~~
/afs/cern.ch/user/b/brfranco/work/public/check_monolithic_dr/k4geo/detector/calorimeter/dual-readout/src/DRconstructor.cpp: In member function 'void ddDRcalo::DRconstructor::implementFibers(xml_comp_t&, dd4hep::Volume&, dd4hep::Trap&, dd4hep::DDSegmentation::DRparamBase_k4geo*, int)':
/afs/cern.ch/user/b/brfranco/work/public/check_monolithic_dr/k4geo/detector/calorimeter/dual-readout/src/DRconstructor.cpp:142:168: warning: unused parameter 'towerNo' [-Wunused-parameter]
  142 | void ddDRcalo::DRconstructor::implementFibers(xml_comp_t& x_theta, dd4hep::Volume& towerVol, dd4hep::Trap& trap, dd4hep::DDSegmentation::DRparamBase_k4geo* param, int towerNo) {
      |                                                                                                                                                                    ~~~~^~~~~~~
/afs/cern.ch/user/b/brfranco/work/public/check_monolithic_dr/k4geo/detector/calorimeter/dual-readout/src/DRconstructor.cpp: In member function 'void ddDRcalo::DRconstructor::implementFiber(dd4hep::Volume&, dd4hep::Trap&, dd4hep::Position, int, int, float)':
/afs/cern.ch/user/b/brfranco/work/public/check_monolithic_dr/k4geo/detector/calorimeter/dual-readout/src/DRconstructor.cpp:237:86: warning: unused parameter 'trap' [-Wunused-parameter]
  237 | void ddDRcalo::DRconstructor::implementFiber(dd4hep::Volume& towerVol, dd4hep::Trap& trap, dd4hep::Position pos, int col, int row, float fiberLen) {
      |                                                                        ~~~~~~~~~~~~~~^~~~
/afs/cern.ch/user/b/brfranco/work/public/check_monolithic_dr/k4geo/detector/calorimeter/dual-readout/src/FiberDualReadoutCalo_o1_v01.cpp: In function 'dd4hep::Ref_t ddDRcalo::create_detector(dd4hep::Detector&, xml_h, dd4hep::SensitiveDetector)':
/afs/cern.ch/user/b/brfranco/work/public/check_monolithic_dr/k4geo/detector/calorimeter/dual-readout/src/FiberDualReadoutCalo_o1_v01.cpp:34:28: warning: variable 'mirrorSurfProp' set but not used [-Wunused-but-set-variable]
   34 |     dd4hep::OpticalSurface mirrorSurfProp = surfMgr.opticalSurface("/world/"+name+"#MirrorSurf");
```

**Comment by BrieucF** [comment]
And actually, the code does not compile: https://github.com/key4hep/k4geo/actions/runs/9784873250/job/27016711979?pr=346#step:3:440

File: plugins/Geant4Output2EDM4hep_DRC.cpp
Comment (swkim95) [reply plugins/Geant4Output2EDM4hep_DRC.cpp:459]: I needed this plugin to store the custom hit class (Geant4DRCalorimeter::Hit).
I always sourced the nightly builds before compiling the k4geo (`source /cvmfs/sw-nightlies.hsf.org/key4hep/setup.sh`) but still this happened.
I was working on the CentOS7 machine (local server), and testing on Alma9(lxplus) solved the errors.

Anyway I updated this code following the latest dd4hep master branch `Geant4Output2EDM4hep.cpp` code.

File: FCCee/IDEA/compact/IDEA_o1_v03/FiberDualReadoutCalo_o1_v01.xml
Comment (SanghyunKo) [inline FCCee/IDEA/compact/IDEA_o1_v03/FiberDualReadoutCalo_o1_v01.xml:589]: seems like CI is failing because we have a customized SD (type `DRcaloSiPMSD`) but the CI still looks for the standard SD (type `calorimeter`). Do we want to update the CI matrix itself here? (since the usage of the standard SD is not intended)
```
DDSim            INFO Setting up SD for DRcalo with No Default
DDSim            ERROR Cannot find Action for detector DRcalo. You have to extend "action.mapAction"
DDSim            ERROR Failed setting up sensitive detector Cannot find Action
```

Comment (andresailer) [reply FCCee/IDEA/compact/IDEA_o1_v03/FiberDualReadoutCalo_o1_v01.xml:589]: Don't you just have to provide a steering file to configure the correct SD?
(is that what you mean by updating the CI matrix?)

Comment (SanghyunKo) [reply FCCee/IDEA/compact/IDEA_o1_v03/FiberDualReadoutCalo_o1_v01.xml:589]: That's exactly what we do when running 'offline' (and it works fine), but I suspect the github CI simply executes `ddsim --compactFile IDEA_o1_v03.xml` (maybe this [one](https://github.com/key4hep/k4geo/blob/main/test/CMakeLists.txt#L94)?) without a steering file and then fails. So I was wondering whether we want to fix this.

Comment (andresailer) [reply FCCee/IDEA/compact/IDEA_o1_v03/FiberDualReadoutCalo_o1_v01.xml:589]: The test must not fail, so we want to fix this.

Comment (SanghyunKo) [reply FCCee/IDEA/compact/IDEA_o1_v03/FiberDualReadoutCalo_o1_v01.xml:589]: I think I figured out how to do this without pushing a steering file. @swkim95 can you try the following for this [line](https://github.com/key4hep/k4geo/blob/main/test/CMakeLists.txt#L94)? `ddsim --compactFile=${CMAKE_CURRENT_SOURCE_DIR}/../FCCee/IDEA/compact/IDEA_o1_v03/IDEA_o1_v03.xml --runType=batch -G -N=1 --outputFile=testIDEA_o1_v03.slcio --action.calorimeterSDTypes='calorimeter DRcaloSiPMSD' --action.mapActions='DRcalo DRCaloSDAction' )`

File: plugins/FiberDRCaloSDAction.cpp
Comment (BrieucF) [inline plugins/FiberDRCaloSDAction.cpp:231]: @andresailer is this what you meant with 

> To ensure reproducibility, you must use the dd4hep random interfaces things.

?

File: plugins/FiberDRCaloSDAction.cpp
Comment (BrieucF) [inline plugins/FiberDRCaloSDAction.cpp:195]:
```suggestion
    G4bool Geant4SensitiveAction<DRCData>::process(G4Step GEANT4_CONST_STEP *step, G4TouchableHistory*)
```

Currently I still see a warning

File: example/SteeringFile_IDEA_o1_v03.py
Comment (BrieucF) [inline example/SteeringFile_IDEA_o1_v03.py:241]: The test would now shoot particles in the beampipe and thus not cross any detector, could you change this?

Comment (SanghyunKo) [reply example/SteeringFile_IDEA_o1_v03.py:241]: @BrieucF interesting, the steering file setup other than the SD and the file I/O part should be identical to other `CTest` setups since he used `--dumpSteeringFile` to configure the steering file. Could this mean that the same thing is happening for other `CTest` as well?

Comment (BrieucF) [reply example/SteeringFile_IDEA_o1_v03.py:241]: Yes, it seems so

Comment (BrieucF) [reply example/SteeringFile_IDEA_o1_v03.py:241]: I guess it matters less since most of the other subdetector uses "central" sensitive actions. Unless the sensitive action are being called anyway at some point, I would make sure we actually have some activity in the dual-readout so that this is also being 'tested'

Comment (swkim95) [reply example/SteeringFile_IDEA_o1_v03.py]: I updated direction to (1., 1., 1.)

File: plugins/FiberDRCaloSDAction.cpp
Comment (BrieucF) [reply plugins/FiberDRCaloSDAction.cpp:231]: I checked, the simulation seems to be reproducible

**Comment by SanghyunKo** [comment]
Final remarks before merging the PR

1) The code lacks comments, especially in the geometry part - this is going to be improved throughout further iterations
2) Regarding the addition of custom Geant4 output action - this is certainly not ideal, but we'd like to keep it temporarily until we get the persistent list of output variables we need downstream. We could probably add `Geant4CalorimeterTimingAction` or something similar to DD4hep in the future (I suspect we need timing but not wavelength information).

**Review by BrieucF — APPROVED** [review]
Approving this now because 
- having the geometry already unlocks some studies on multiple scattering
- having it in the central repository makes sure it will be tested from now on and makes life easier for people needing the detector
- the sensitive action will be improved in subsequent PR

**Comment by BrieucF** [comment]
Can we merge this?

**Comment by andresailer** [comment]
There are conflicts...

**Comment by BrieucF** [comment]
With what?

**Comment by andresailer** [comment]
image

File: FCCee/IDEA/compact/IDEA_o1_v03/DectDimensions_IDEA_o1_v03.xml
Comment (andresailer) [inline FCCee/IDEA/compact/IDEA_o1_v03/DectDimensions_IDEA_o1_v03.xml:59]: Discussed with @BrieucF to have Calo ID lower than muonID

**Pull request merged**
